controller/destination/endpoints_watcher.go

```diff
@@ -346,26 +334,39 @@ func (sp *servicePort) updateService(newService *v1.Service) {
 		}
 	}
 	if newTargetPort != sp.targetPort {
-		newAddresses := addresses(sp.endpoints, newTargetPort)
-		sp.updateAddresses(newAddresses)
+		sp.updateAddresses(sp.endpoints, newTargetPort)
 		sp.targetPort = newTargetPort
 	}
 }
 
-func (sp *servicePort) updateAddresses(newAddresses []common.TcpAddress) {
+func (sp *servicePort) updateAddresses(endpoints *v1.Endpoints, port intstr.IntOrString) {
+	addresses := sp.addressMap(endpoints, port)
+	newAddresses := make([]common.TcpAddress, 0)
+	for a := range addresses {
+		newAddresses = append(newAddresses, a)
+	}
+
 	log.Debugf("Updating %s:%d to %s", sp.service, sp.port, addr.AddressesToString(newAddresses))
 
 	if len(newAddresses) == 0 {
 		for _, listener := range sp.listeners {
 			listener.NoEndpoints(true)
 		}
 	} else {
-		add, remove := addr.DiffAddresses(sp.addresses, newAddresses)
+		oldAddresses := make([]common.TcpAddress, 0)
+		for a := range sp.addresses {
+			oldAddresses = append(oldAddresses, a)
+		}
+		add, remove := addr.DiffAddresses(oldAddresses, newAddresses)
+		addAddresses := make(map[common.TcpAddress]*v1.Pod)
+		for _, a := range add {
+			addAddresses[a] = addresses[a]
+		}
 		for _, listener := range sp.listeners {
-			listener.Update(add, remove)
+			listener.Update(addAddresses, remove)
 		}
 	}
-	sp.addresses = newAddresses
+	sp.addresses = addresses
```

Comment:
also realizing this code was already here, but is it possible for `updateAddresses` to get called by two different threads? does `sp.addresses` need mutex protection here? i see the mutex used in `updateEndpoints` and `updateService` but is it safer / more explicit to use the mutex at this lower-level?

Reply:
Good catch! Yeah, the mutex handling on this whole struct is a bit messy. I think it deserve a larger refactor, but I'm not going to do that as part of this branch. As it's currently implemented, at least the mutex is already locked whenever this method is called.